### PR TITLE
Sum dtype

### PR DESCRIPTION
**Pull request by unnonouno**
sum supports all float types.
related to #555

### CODE REVIEW

File: tests/chainer_tests/functions_tests/math_tests/test_sum.py [hunk 1]
```diff
@@ -11,18 +11,29 @@
 from chainer.testing import condition
 
 
+@testing.parameterize(
+    {'dtype': numpy.float16},
+    {'dtype': numpy.float32},
+    {'dtype': numpy.float64},
+)
```
Comment (rezoo) [inline tests/chainer_tests/functions_tests/math_tests/test_sum.py]: I found that this class only tests the tensor of order three. How about using ``testing.product()`` and writing the following?
```python
@testing.parameterize(*testing.product({
    'shape': [(3, 2, 4), (2, 3, 1, 4)],
    'dtype': [numpy.float16, numpy.float32, numpy.float64],
}))
```

File: tests/chainer_tests/functions_tests/math_tests/test_sum.py [hunk 2]
```diff
@@ -11,18 +11,29 @@
...
 class TestSum(unittest.TestCase):
 
     def setUp(self):
-        self.x = numpy.random.uniform(-1, 1, (3, 2, 4)).astype(numpy.float32)
-        self.gy = numpy.random.uniform(-1, 1, ()).astype(numpy.float32)
+        self.x = numpy.random.uniform(-1, 1, (3, 2, 4)).astype(self.dtype)
```
Comment (rezoo) [inline tests/chainer_tests/functions_tests/math_tests/test_sum.py]: If you agree that adding more tests regarding ``shape``, this line should be changed like
```python
self.x = numpy.random.uniform(-1, 1, (3, 2, 4)).astype(self.shape)
```

File: tests/chainer_tests/functions_tests/math_tests/test_sum.py (hunk 1 above)
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/math_tests/test_sum.py]: OK, i'll fix to use `product`.
Your suggestion is Ok, but I think that is another issue. I'll send another PR about shapes.

**Comment by rezoo** [comment]
Great, then I will merge it.

**Pull request merged**